### ISSUE TITLE
snakemake --gui "TypeError: 'function' object is not iterable"

### ISSUE THREAD

**Issue by millerh1** [snakemake/snakemake#470]
**Snakemake version**
5.19.3

**Describe the bug**
When attempting to use `snakemake --gui` an error is return.

**Logs**
Error output: 

```
Traceback (most recent call last):
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/bin/snakemake", line 10, in <module>
    sys.exit(main())
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/__init__.py", line 2195, in main
    gui.register(_snakemake, args)
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/gui.py", line 61, in register
    run_snakemake(list_target_rules=True, log_handler=log_handler)
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/__init__.py", line 415, in snakemake
    setup_logger(
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/logging.py", line 460, in setup_logger
    logger.log_handler.extend(handler)
TypeError: 'function' object is not iterable

```

**Minimal example**
Add a minimal example for reproducing the bug.

`Snakefile`:

```
# Make the file
f = open('file.txt', "w+")
f.writelines("Hello world")
f.close()

rule all:
    input: "file.txt"
    output: "newfile.txt"
    shell: "cat {input} > {output}"
```

command line and output:

```
(mamba_env) millerh1@cbbi16:~/Bishop.lab/Projects/RSeq$ snakemake -np
Building DAG of jobs...
Job counts:
        count   jobs
        1       all
        1

[Tue Jun 30 15:37:16 2020]
rule all:
    input: file.txt
    output: newfile.txt
    jobid: 0

cat file.txt > newfile.txt
Job counts:
        count   jobs
        1       all
        1
This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
(mamba_env) millerh1@cbbi16:~/Bishop.lab/Projects/RSeq$ snakemake --gui --cores 1
Traceback (most recent call last):
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/bin/snakemake", line 10, in <module>
    sys.exit(main())
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/__init__.py", line 2195, in main
    gui.register(_snakemake, args)
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/gui.py", line 61, in register
    run_snakemake(list_target_rules=True, log_handler=log_handler)
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/__init__.py", line 415, in snakemake
    setup_logger(
  File "/home/UTHSCSA/millerh1/miniconda3/envs/mamba_env/lib/python3.8/site-packages/snakemake/logging.py", line 460, in setup_logger
    logger.log_handler.extend(handler)
TypeError: 'function' object is not iterable

```

**Additional context**

**Comment by millerh1** [comment]
`snakemake.snakemake` needs log_handler provided as a list, not a function. 

with `example.smk`:

```
rule all:
    output: 'helloworld.txt'
    shell: 'echo "hello world!" > {output}'
```

This produces an error:

```
def log_handler(msg):
    print(msg)

snk.snakemake('example.smk', log_handler=log_handler)
```

error:

```
Traceback (most recent call last):
  File "<input>", line 4, in <module>
  File "/home/UTHSCSA/millerh1/miniconda3/envs/RSeqEnv/lib/python3.7/site-packages/snakemake/__init__.py", line 426, in snakemake
    show_failed_logs=show_failed_logs,
  File "/home/UTHSCSA/millerh1/miniconda3/envs/RSeqEnv/lib/python3.7/site-packages/snakemake/logging.py", line 472, in setup_logger
    logger.log_handler.extend(handler)
TypeError: 'function' object is not iterable

```

This works:

```
def log_handler(msg):
    print(msg)

snk.snakemake('example.smk', log_handler=[log_handler])
```
output:

```
Building DAG of jobs...
{'level': 'info', 'msg': 'Building DAG of jobs...', 'indent': False}
{'level': 'info', 'msg': 'Building DAG of jobs...', 'indent': False}
{'level': 'info', 'msg': 'Building DAG of jobs...', 'indent': False}
{'level': 'dag_debug', 'status': 'candidate', 'job': all}
{'level': 'dag_debug', 'status': 'candidate', 'job': all}
{'level': 'dag_debug', 'status': 'candidate', 'job': all}
{'level': 'dag_debug', 'status': 'selected', 'job': all}
{'level': 'dag_debug', 'status': 'selected', 'job': all}
{'level': 'dag_debug', 'status': 'selected', 'job': all}
{'level': 'dag_debug', 'file': None, 'msg': 'Producer found, hence exceptions are ignored.', 'exception': WorkflowError('')}
{'level': 'dag_debug', 'file': None, 'msg': 'Producer found, hence exceptions are ignored.', 'exception': WorkflowError('')}
{'level': 'dag_debug', 'file': None, 'msg': 'Producer found, hence exceptions are ignored.', 'exception': WorkflowError('')}
Using shell: /bin/bash
{'level': 'info', 'msg': 'Using shell: /bin/bash', 'indent': False}
Provided cores: 1 (use --cores to define parallelism)
{'level': 'info', 'msg': 'Using shell: /bin/bash', 'indent': False}
{'level': 'info', 'msg': 'Using shell: /bin/bash', 'indent': False}
{'level': 'resources_info', 'msg': 'Provided cores: 1 (use --cores to define parallelism)'}
{'level': 'resources_info', 'msg': 'Provided cores: 1 (use --cores to define parallelism)'}
{'level': 'resources_info', 'msg': 'Provided cores: 1 (use --cores to define parallelism)'}
Rules claiming more threads will be scaled down.
{'level': 'resources_info', 'msg': 'Rules claiming more threads will be scaled down.'}
{'level': 'resources_info', 'msg': 'Rules claiming more threads will be scaled down.'}
{'level': 'resources_info', 'msg': 'Rules claiming more threads will be scaled down.'}
Job counts:
	count	jobs
	1	all
	1
{'level': 'run_info', 'msg': 'Job counts:\n\tcount\tjobs\n\t1\tall\n\t1'}
{'level': 'run_info', 'msg': 'Job counts:\n\tcount\tjobs\n\t1\tall\n\t1'}
{'level': 'run_info', 'msg': 'Job counts:\n\tcount\tjobs\n\t1\tall\n\t1'}
{'level': 'debug', 'msg': "Resources before job selection: {'_cores': 1, '_nodes': 9223372036854775807}"}
{'level': 'debug', 'msg': "Resources before job selection: {'_cores': 1, '_nodes': 9223372036854775807}"}
{'level': 'debug', 'msg': "Resources before job selection: {'_cores': 1, '_nodes': 9223372036854775807}"}
{'level': 'debug', 'msg': 'Ready jobs (1):\n\tall'}
{'level': 'debug', 'msg': 'Ready jobs (1):\n\tall'}
{'level': 'debug', 'msg': 'Ready jobs (1):\n\tall'}
{'level': 'debug', 'msg': 'Selected jobs (1):\n\tall'}
{'level': 'debug', 'msg': 'Selected jobs (1):\n\tall'}
{'level': 'debug', 'msg': 'Selected jobs (1):\n\tall'}
{'level': 'debug', 'msg': "Resources after job selection: {'_cores': 0, '_nodes': 9223372036854775806}"}
{'level': 'debug', 'msg': "Resources after job selection: {'_cores': 0, '_nodes': 9223372036854775806}"}
{'level': 'debug', 'msg': "Resources after job selection: {'_cores': 0, '_nodes': 9223372036854775806}"}
[Tue Jul 14 08:27:47 2020]
rule all:
    output: helloworld.txt
    jobid: 0
{'jobid': 0, 'msg': None, 'name': 'all', 'local': False, 'input': [], 'output': ['helloworld.txt'], 'log': [], 'benchmark': None, 'wildcards': {}, 'reason': 'Missing output files: helloworld.txt', 'resources': [1, 1], 'priority': 0, 'threads': 1, 'indent': False, 'is_checkpoint': False, 'printshellcmd': True, 'level': 'job_info'}
{'jobid': 0, 'msg': None, 'name': 'all', 'local': False, 'input': [], 'output': ['helloworld.txt'], 'log': [], 'benchmark': None, 'wildcards': {}, 'reason': 'Missing output files: helloworld.txt', 'resources': [1, 1], 'priority': 0, 'threads': 1, 'indent': False, 'is_checkpoint': False, 'printshellcmd': True, 'level': 'job_info'}
{'jobid': 0, 'msg': None, 'name': 'all', 'local': False, 'input': [], 'output': ['helloworld.txt'], 'log': [], 'benchmark': None, 'wildcards': {}, 'reason': 'Missing output files: helloworld.txt', 'resources': [1, 1], 'priority': 0, 'threads': 1, 'indent': False, 'is_checkpoint': False, 'printshellcmd': True, 'level': 'job_info'}
{'level': 'shellcmd', 'msg': 'echo "hello world!" > helloworld.txt', 'indent': False}
{'level': 'shellcmd', 'msg': 'echo "hello world!" > helloworld.txt', 'indent': False}
{'level': 'shellcmd', 'msg': 'echo "hello world!" > helloworld.txt', 'indent': False}
[Tue Jul 14 08:27:47 2020]
Finished job 0.
{'jobid': 0, 'level': 'job_finished'}
{'jobid': 0, 'level': 'job_finished'}
{'jobid': 0, 'level': 'job_finished'}
1 of 1 steps (100%) done
{'level': 'progress', 'done': 1, 'total': 1}
{'level': 'progress', 'done': 1, 'total': 1}
{'level': 'progress', 'done': 1, 'total': 1}
Complete log: /home/UTHSCSA/millerh1/Bishop.lab/Projects/RSeq/.snakemake/log/2020-07-14T082747.326926.snakemake.log
{'level': 'info', 'msg': 'Complete log: /home/UTHSCSA/millerh1/Bishop.lab/Projects/RSeq/.snakemake/log/2020-07-14T082747.326926.snakemake.log', 'indent': False}
{'level': 'info', 'msg': 'Complete log: /home/UTHSCSA/millerh1/Bishop.lab/Projects/RSeq/.snakemake/log/2020-07-14T082747.326926.snakemake.log', 'indent': False}
{'level': 'info', 'msg': 'Complete log: /home/UTHSCSA/millerh1/Bishop.lab/Projects/RSeq/.snakemake/log/2020-07-14T082747.326926.snakemake.log', 'indent': False}
{'level': 'debug', 'msg': 'unlocking'}
{'level': 'debug', 'msg': 'unlocking'}
{'level': 'debug', 'msg': 'unlocking'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removing lock'}
{'level': 'debug', 'msg': 'removed all locks'}
{'level': 'debug', 'msg': 'removed all locks'}
{'level': 'debug', 'msg': 'removed all locks'}
True

```

**Comment by johanneskoester** [comment]
Thank you guys. The gui should now work again, although it is still basic and experimental.

**Comment by johanneskoester** [comment]
Fixed in commit 3013cb36.
